src/nupic/research/temporal_memory.py

```diff
@@ -120,419 +209,192 @@ def compute(self, activeColumns, learn=True):
-                                           predictedActiveColumns,
-                                           prevActiveCells,
-                                           prevWinnerCells,
-                                           self.connections)
-
-    activeCells.update(_activeCells)
-    winnerCells.update(_winnerCells)
-
-    if learn:
-      self.learnOnSegments(prevActiveSegments,
-                           learningSegments,
-                           prevActiveCells,
-                           winnerCells,
-                           prevWinnerCells,
-                           self.connections,
-                           predictedInactiveCells,
-                           prevMatchingSegments)
+
+    activeColumns = sorted(activeColumns)
+
+    self.activeCells = []
+    self.winnerCells = []
+
+    for excitedColumn in excitedColumnsGenerator(activeColumns,
+                                                 self.activeSegments,
+                                                 self.matchingSegments,
+                                                 self.cellsPerColumn,
+                                                 self.connections):
+      if excitedColumn["isActiveColumn"]:
+        if len(excitedColumn["activeSegments"]) != 0:
+          cellsToAdd = self.activatePredictedColumn(excitedColumn, learn,
+                                                    prevActiveCells)
+          self.activeCells += cellsToAdd
+          self.winnerCells += cellsToAdd
+        else:
+          (cellsToAdd,
+           winnerCell) = self.burstColumn(excitedColumn, learn,
+                                          prevActiveCells, prevWinnerCells)
+          self.activeCells += cellsToAdd
+          self.winnerCells.append(winnerCell)
+      else:
+        if learn:
+          self.punishPredictedColumn(excitedColumn, prevActiveCells)
 
     (activeSegments,
-     predictiveCells,
-     matchingSegments,
-     matchingCells) = self.computePredictiveCells(activeCells, self.connections)
+     matchingSegments) = self.connections.computeActivity(self.activeCells,
+                                                          self.connectedPermanence,
+                                                          self.activationThreshold,
+                                                          0.0,
+                                                          self.minThreshold)
 
-    self.activeCells = activeCells
-    self.winnerCells = winnerCells
     self.activeSegments = activeSegments
-    self.predictiveCells = predictiveCells
     self.matchingSegments = matchingSegments
-    self.matchingCells = matchingCells
-
 
   def reset(self):
     """
     Indicates the start of a new sequence. Resets sequence state of the TM.
     """
-    self.activeCells = set()
-    self.predictiveCells = set()
-    self.activeSegments = set()
-    self.winnerCells = set()
-
+    self.activeCells = []
+    self.activeSegments = []
+    self.winnerCells = []
```

Comment:
This should also reset the `matchingSegments`.

(I realize the code had this issue before. It might not have caused problems because the old code doesn't make full use of the matchingSegments, I think it only used them for punishment but not for computing the best matching segment -- it essentially computed the matching segments twice.)